### DISCUSSION TITLE
Static infinite plane shape

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#465] [Ideas]
I'm currently in the process of implementing the last few items of my Godot extension needed to bring it somewhat up-to-par with Godot Physics. This includes implementing two custom shapes, called [`SeparationRayShape3D`](https://docs.godotengine.org/en/4.0/classes/class_separationrayshape3d.html) and [`WorldBoundaryShape3D`](https://docs.godotengine.org/en/4.0/classes/class_worldboundaryshape3d.html).

The `SeparationRayShape3D` shape is effectively just a ray-cast implemented as a convex shape, albeit with some quirky characteristics like altering the collision normal based on its `slide_on_slope` property. Either way, I think I have that one under control, and it seems somewhat special-purpose, so having it be a custom shape seems appropriate.

The `WorldBoundaryShape3D` shape however, is nothing more than a static infinite plane. Its somewhat unusual name apparently comes from wanting to keep it consistent with the infinite line shape that they have for 2D, which is appropriately called `WorldBoundaryShape2D`. I figured this would be somewhat easy to implement, but I must admit that I'm running up against the limits of my physics/math abilities.

I did manage to replicate what Bullet does in [its implementation](https://github.com/bulletphysics/bullet3/blob/0e59474e1cf1fd69e3ca512826cfbe15cd6a9cec/src/BulletCollision/CollisionShapes/btStaticPlaneShape.cpp#L31-L75), by making it a concave shape consisting of two very large triangles, and relying on `CollideConvexVsTriangles` in the collide shape function. Although, while this did work, it didn't take much in terms of size to either bring out what I presume were precision issues, in the form of spheres "skipping" while rolling down one of these planes, or to crash due to NaNs.

With all that said, I was wondering if a plane shape like this is something you've considered adding to Jolt in the past, or might be willing to add in the future, assuming you see some way of implementing it that wouldn't suffer from the issues described above? From what I can tell this shape type is something PhysX and Havok offers as well, so it seems like something that others might find useful.

**Comment by jrouwe** [comment]
Hello,

Please have a look at this issue: https://github.com/jrouwe/JoltPhysics/issues/126

**Reply by mihe** [reply]
Thank you for the link. I should've figured someone would have asked this already.

This will have to be one of the things I won't support then. The broad phase problem, which I hadn't considered, sounds like it might lead to performance issues that are not obvious and hard to track down, and subdividing the plane to the point where the triangles are more sensible feels like it would defeat the assumed simplicity of the shape.

Oh well. :)

Thank you for the quick reply!
